codec.go

```diff
@@ -34,38 +34,102 @@ func (e *DecodeProtocolMessageFieldError) Error() string {
 	return fmt.Sprintf("%s %s '%s' is not supported (seq: %d)", e.SubType, e.FieldName, e.FieldValue, e.Seq)
 }
 
-// DecodeProtocolMessage parses the JSON-encoded data and returns the result of
+// defaultCodec is used to decode vanilla DAP messages.
+var defaultCodec = NewCodec()
+
+// Codec is responsible for turning byte blobs into DAP messages.
+type Codec struct {
+	eventCtor    map[string]messageCtor
+	requestCtor  map[string]messageCtor
+	responseCtor map[string]messageCtor
+}
+
+// NewCodec constructs a new codec that extends the vanilla DAP protocol.
+// Unless you need to register custom DAP messages, use
+// DecodeProtocolMessage instead.
+func NewCodec() *Codec {
+	ret := &Codec{
+		eventCtor:    make(map[string]messageCtor),
+		requestCtor:  make(map[string]messageCtor),
+		responseCtor: make(map[string]messageCtor),
+	}
+	for v, k := range eventCtor {
```

Comment:
Could you swap k and v here and below please?

Reply:
Oops :)
Done.

Reply:
@suzmue PTAL. I forgot to add a test initially. Now I added one.

There wasn't a perfect place to define the custom messages for testing. PLMK if I should move this code.